lib/vacuum/prune/pack/pack.go

```diff
@@ -150,124 +140,219 @@ func (r *cleanup) mark() (required packageMap, err error) {
+// withDependencies computes depender packages for the specified package pkg using
+// the specified depender search algorithm and the existing package state
+func (r *cleanup) withDependencies(pkg existingPackage, depender dependerFunc, state map[loc.Locator]statePackage) (items []statePackage, err error) {
+	dependerPackages, err := depender(pkg.PackageEnvelope, r.Packages)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
+		r.Warnf("Orphaned package %v.", pkg.Locator)
+		return []statePackage{
+			statePackage{existingPackage: pkg},
+		}, nil
 	}
-	for packageFilter, pkg := range packages {
-		if parentPackageFilter.IsEqualTo(packageFilter) {
-			return &pkg, nil
+
+	for _, dependerPackage := range dependerPackages {
+		dependerVersion, err := dependerPackage.Locator.SemVer()
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		var parent statePackage
+		var exists bool
+		if parent, exists = state[dependerPackage.Locator]; !exists {
+			parent = statePackage{
+				existingPackage: existingPackage{
+					PackageEnvelope: dependerPackage,
+					Version:         *dependerVersion,
+				},
+			}
 		}
+		parent.dependencies = append(parent.dependencies, statePackage{existingPackage: pkg})
+		items = append(items, parent)
 	}
-	return nil, trace.NotFound("no package found for configuration package %v", envelope.Locator)
+	return items, nil
 }
 
-func packageForResources(envelope pack.PackageEnvelope, packages packageMap) (*requiredPackage, error) {
+// dependerFunc computes a depender package for specified package
+type dependerFunc func(pack.PackageEnvelope, packageService) ([]pack.PackageEnvelope, error)
+
+func packageForConfig(envelope pack.PackageEnvelope, service packageService) ([]pack.PackageEnvelope, error) {
+	parentPackageRef := envelope.RuntimeLabels[pack.ConfigLabel]
+	parentPackageFilter, err := loc.ParseLocator(parentPackageRef)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	parentPackage := loc.Locator{
+		Repository: parentPackageFilter.Repository,
+		Name:       parentPackageFilter.Name,
+		Version:    envelope.Locator.Version,
+	}
+
+	return dependerForPackage(envelope, service, parentPackage)
+}
+
+func packageForResources(envelope pack.PackageEnvelope, service packageService) ([]pack.PackageEnvelope, error) {
 	appPackageName := strings.TrimSuffix(envelope.Locator.Name, "-resources")
-	appPackageFilter := loc.Locator{
+	appPackage := loc.Locator{
 		Repository: envelope.Locator.Repository,
 		Name:       appPackageName,
-		Version:    loc.ZeroVersion,
+		Version:    envelope.Locator.Version,
 	}
-	for packageFilter, pkg := range packages {
-		if appPackageFilter.IsEqualTo(packageFilter) {
-			return &pkg, nil
+	return dependerForPackage(envelope, service, appPackage)
+}
+
+func dependerForPackage(envelope pack.PackageEnvelope, service packageService, depender loc.Locator) ([]pack.PackageEnvelope, error) {
```

Comment:
Does this function need both `envelope` and `depender`? Isn't `depender` the same locator as `envelope.Package`?

Reply:
See above: depender is the owning package for the package described by `envelope`. It feels that a better comment is in order.